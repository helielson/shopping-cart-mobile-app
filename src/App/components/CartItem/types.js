/**
 * Types definition used on CartItem Component
 *
 * @format
 * @flow
 */
import type { itemType } from '../../flow/types';

// PropTypes for CloseButton
export type CloseButtonProps = {
  onPress: () => mixed,
};

// PropTypes for CartItem
export type CartItemProps = {
  item: itemType,
  addOne: (item: itemType) => mixed,
  removeOne: (item: itemType) => mixed,
  removeItem: (id: string) => void,
};
