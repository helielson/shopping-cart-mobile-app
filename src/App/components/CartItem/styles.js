import colors from '../../config/colors';

const shadowStyle = {
  shadowOpacity: 0.35,
  shadowOffset: {
    width: 0,
    height: 1,
  },
  shadowColor: '#000',
  shadowRadius: 2,
  elevation: 3,
};

export default {
  container: {
    alignItems: 'center',
    marginTop: 15,
    marginHorizontal: 15,
    backgroundColor: colors.backgroundLight,
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    ...shadowStyle,
  },
  pictureContainer: {
    padding: 10,
  },
  picture: {
    height: 110,
    width: 110,
    backgroundColor: colors.background,
    borderWidth: 2,
    borderColor: colors.background,
  },
  leftContainer: {
    flexDirection: 'column',
    flex: 1,
  },
  content: {
    flexDirection: 'row',
  },
  titleContainer: {
    flex: 1,
    justifyContent: 'center',
  },
  middleContainer: {
    flex: 2,
    flexDirection: 'column',
    justifyContent: 'space-around',
  },
  productTitle: {
    fontSize: 17,
    fontWeight: 'bold',
    color: '#333',
  },
  buttonsContainer: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'center',
  },
  quantityButton: {
    padding: 5,
  },
  quantity: {
    color: colors.ternary,
    fontWeight: 'bold',
  },
  moneyContainer: {
    flexDirection: 'column',
    justifyContent: 'space-between',
    flex: 1,
  },
  price: {
    color: colors.primary,
    flex: 1,
  },
  total: {
    color: colors.ternary,
    flex: 1,
  },
};
