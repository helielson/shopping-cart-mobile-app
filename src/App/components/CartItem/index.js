/**
 * @format
 * @flow
 */
import React from 'react';
import { connect } from 'react-redux';
import { View, Text, Image, TouchableOpacity, Alert } from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import type { CloseButtonProps, CartItemProps } from './types';
import { addOneItem, removeOneItem, removeFromCart } from '../../store/actions';
import formatCurrency from '../../util';
import styles from './styles';
import colors from '../../config/colors';

const CloseButton = ({ onPress }: CloseButtonProps) => (
  <TouchableOpacity
    hitSlop={{ top: 10, left: 10, right: 10, bottom: 10 }}
    activeOpacity={0.6}
    onPress={onPress}
    style={styles.closeButton}
  >
    <Icon name="close" size={15} color={colors.backgroundLight} />
  </TouchableOpacity>
);

/**
 * Show a Alert to the user that ask if he really want to remove a item
 *
 * @param {(id: string) => void} removeItem Callback to remove the item
 * @param {string} id The id of the product to be removed from cart
 */
const confirmModal = (removeItem: (id: string) => void, id: string) => {
  Alert.alert('Remove Item', 'Are you sure you want to remove this item?', [
    { text: 'Cancel', style: 'cancel' },
    {
      text: 'OK',
      onPress: () => {
        removeItem(id);
      },
    },
  ]);
};

/**
 * CartItem component that allows user to increase or decrease the quantity
 * of a specific item on it cart
 *
 * @param {CartItemProps} {
 *   item,
 *   addOne,
 *   removeOne,
 *   removeItem,
 * }
 * @returns <CartItem/>
 */
const CartItem = (props: CartItemProps) => {
  const { item, addOne, removeOne, removeItem } = props;
  const {
    item: { picture, price, title, onCart = 0, id },
  } = props;
  return (
    <View style={styles.container}>
      <CloseButton onPress={() => confirmModal(removeItem, id)} />
      <View style={styles.pictureContainer}>
        <Image style={styles.picture} source={{ uri: picture }} />
      </View>
      <View style={styles.leftContainer}>
        <View style={styles.titleContainer}>
          <Text style={styles.productTitle} numberOfLines={1} ellipsizeMode="tail">
            {title}
          </Text>
        </View>
        <View style={styles.content}>
          <View style={styles.middleContainer}>
            <View style={styles.moneyContainer}>
              <Text style={styles.price}>{formatCurrency(price)}</Text>
              <Text style={styles.total}>
                <Text style={styles.productTitle}>Total: </Text>
                {formatCurrency(price * onCart)}
              </Text>
            </View>
          </View>
          <View style={styles.buttonsContainer}>
            <TouchableOpacity style={styles.quantityButton} onPress={() => addOne(item)}>
              <Icon name="expand-less" size={30} color={colors.buttonColor} />
            </TouchableOpacity>
            <Text style={styles.quantity}>{onCart}</Text>
            <TouchableOpacity
              disabled={onCart === 1}
              style={styles.quantityButton}
              onPress={() => {
                removeOne(item);
              }}
            >
              <Icon name="expand-more" size={30} color={colors.buttonColor} />
            </TouchableOpacity>
          </View>
        </View>
      </View>
    </View>
  );
};

// Map Redux Actions Dispatchers to Component Props
const mapDispatchToProps = dispatch => ({
  addOne: item => dispatch(addOneItem(item)),
  removeOne: item => dispatch(removeOneItem(item)),
  removeItem: id => dispatch(removeFromCart(id)),
});

// Export Stateless Components (To be used on Tests)
export { CartItem, CloseButton };

// Connect component with store and export it as default
export default connect(
  null,
  mapDispatchToProps
)(CartItem);
