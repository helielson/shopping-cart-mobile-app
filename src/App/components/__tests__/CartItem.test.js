/**
 * @flow
 * @format
 * */
import React from 'react';
import { shallow } from 'enzyme';

import { CartItem } from '../CartItem';

const item = {
  id: 'a',
  description:
    'Análise Sensorial: \n Análise Visual: cor rosa com reflexos alaranjados. Análise Olfativa: Frutas vermelhas (morango) Análise Gustativa: seco, corpo médio, excelente acidez, taninos de boa qualidade, com retro-olfato de frutas.',
  price: 200,
  onCart: 1,
  available: 20,
  picture: 'https://cdn.awsli.com.br/600x700/365/365010/produto/13341816/fe9a59131b.jpg',
  title: 'Vinho Nederburg Rosé',
};

describe('<CartItem />', () => {
  test('should render properly', () => {
    const wrapper = shallow(
      <CartItem item={item} addOne={() => {}} removeOne={() => {}} removeItem={() => {}} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
