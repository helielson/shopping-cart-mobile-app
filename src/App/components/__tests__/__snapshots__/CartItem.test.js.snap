// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> should render properly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(245, 245, 246, 1.00)",
      "elevation": 3,
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginBottom": 5,
      "marginHorizontal": 15,
      "marginTop": 15,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.35,
      "shadowRadius": 2,
    }
  }
>
  <CloseButton
    onPress={[Function]}
  />
  <View
    style={
      Object {
        "padding": 10,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://cdn.awsli.com.br/600x700/365/365010/produto/13341816/fe9a59131b.jpg",
        }
      }
      style={
        Object {
          "backgroundColor": "rgba(225, 226, 225, 1.00)",
          "borderColor": "rgba(225, 226, 225, 1.00)",
          "borderWidth": 2,
          "height": 110,
          "width": 110,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "color": "#333",
            "fontSize": 17,
            "fontWeight": "bold",
          }
        }
      >
        Vinho Nederburg Ros√©
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 2,
            "flexDirection": "column",
            "justifyContent": "space-around",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "rgba(114, 171, 65, 1.00)",
                "flex": 1,
              }
            }
          >
            R$ 200,00
          </Text>
          <Text
            style={
              Object {
                "color": "rgba(59, 129, 181, 1.00)",
                "flex": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#333",
                  "fontSize": 17,
                  "fontWeight": "bold",
                }
              }
            >
              Total: 
            </Text>
            R$ 200,00
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
          }
        }
      >
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
          style={
            Object {
              "padding": 5,
            }
          }
        >
          <Icon
            allowFontScaling={false}
            color="rgba(0, 145, 234, 1.00)"
            name="expand-less"
            size={30}
          />
        </TouchableOpacity>
        <Text
          style={
            Object {
              "color": "rgba(59, 129, 181, 1.00)",
              "fontWeight": "bold",
            }
          }
        >
          1
        </Text>
        <TouchableOpacity
          activeOpacity={0.2}
          disabled={true}
          onPress={[Function]}
          style={
            Object {
              "padding": 5,
            }
          }
        >
          <Icon
            allowFontScaling={false}
            color="rgba(0, 145, 234, 1.00)"
            name="expand-more"
            size={30}
          />
        </TouchableOpacity>
      </View>
    </View>
  </View>
</View>
`;
